### ISSUE TITLE
Invalid Dhall.Decoder based on order of ADT constructors

### ISSUE THREAD

**Issue by SiriusStarr** [dhall-lang/dhall-haskell#1711]
Attempting to import the following Dhall:
```
λ(Recursive : Type)
    → let RecursiveF =
            < AF : Recursive
            | BF
            | CF
            | DF
            | WorksIfFirstF :  Text
            | WorksIfSecondF : Text
            | EF
            >
      in    λ(Fix : RecursiveF → Recursive)
          → Fix (RecursiveF.WorksIfSecondF "" )
```
into the Haskell type:
```
data Recursive
  = A Recursive
  | B
  | C
  | D
  | WorksIfSecond NonemptyText
  | WorksIfFirst NonemptyText
  | E
  deriving (Generic, Show)
```
produces the error:
```
Error: Invalid Dhall.Decoder                                               
                                                                                
Every Decoder must provide an extract function that succeeds if an expression   
matches the expected type.  You provided a Decoder that disobeys this contract  
                                                                                
The Decoder provided has the expected dhall type:                               
                                                                                
↳ < WorksIfFirstF : Text | EF >
                                                                                
and it couldn't extract a value from the well-typed expression:                 
                                                                                
↳ < AF : _@1
  | BF
  | CF
  | DF
  | EF
  | WorksIfFirstF : Text
  | WorksIfSecondF : Text
  >.WorksIfSecondF
    ""
```

This error does not occur if any of `A`, `B`, `C`, `D`, or `E` are removed from both the Haskell and Dhall, or if `WorksIfFirst` and `WorksIfSecond` are flipped in the Haskell.  In any of these cases, the expected "failed extraction" error occurs (as intended).

Relevant Haskell instances below:
```
newtype NonemptyText = NonemptyText {getNonemptyText :: Text} deriving (Eq, Ord, Show)

validateText :: Decoder Text -> Decoder NonemptyText
validateText (Decoder extractIn expectedIn) = Decoder extractOut expectedIn
  where
    extractOut e = case extractIn e of
      Success t ->
        if T.null t
          then extractError "Text cannot be empty."
          else Success $ NonemptyText t
      Failure f -> Failure f

instance FromDhall NonemptyText where
  autoWith _ = validateText strictText

data Recursive
  = A Recursive
  | B
  | C
  | D
  | WorksIfFirst NonemptyText
  | WorksIfSecond NonemptyText
  | E
  deriving (Generic, Show)

TH.makeBaseFunctor ''Recursive

deriving instance Generic (RecursiveF a)

deriving instance FromDhall a => FromDhall (RecursiveF a)

convertRecursive :: Fix RecursiveF -> Recursive
convertRecursive = Fix.cata Foldable.embed

instance FromDhall Recursive where
  autoWith opts = convertRecursive <$> autoWith opts
```

Full source of the minimal example (or as minimal as I could make it) is in a [gist here.](https://gist.github.com/SiriusStarr/f42f36781e7f2f39b478c2ab8a6c018c)

**Comment by akrmn** [comment]
Hi @SiriusStarr, 

I was able to shrink your example a bit:

```haskell
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}

module Main where

import Data.Set
import Dhall
import Text.Heredoc

data Foo
  = A (Set ())
  | B
  | C
  | D
  deriving (Generic, Show)

instance FromDhall Foo

main :: IO ()
main = do
  result <-
    Dhall.input
      Dhall.auto
      [here|
let Foo =
      < B
      | C
      | D
      | A : List {}
      >
in Foo.A ([{=}, {=}] : List {})
|] ::
      IO Foo
  print result
```

This means that we can rule out recursive data types, Data.Fix and TemplateHaskell as the reason for this issue.

The error message we're currently getting is a bit confusing, mainly because the expected type mentioned doesn't match the Haskell type. I believe this might be due to the way some of the `GenericFromDhall` instances for `(:+:)` are written. In particular, these lines of `Dhall` ([line 1482](https://github.com/dhall-lang/dhall-haskell/blob/0c4b59a02591b65f2076b801e9b5a11d2b2e5596/dhall/src/Dhall.hs#L1482), [line 1545](https://github.com/dhall-lang/dhall-haskell/blob/0c4b59a02591b65f2076b801e9b5a11d2b2e5596/dhall/src/Dhall.hs#L1545) seem related. I think the `GenericFromDhall` instances for `(:+:)` need to be rewritten in a way that doesn't check unused branches of the union.

**Comment by Gabriel439** [comment]
I have a partial fix up here: https://github.com/dhall-lang/dhall-haskell/pull/1720

**Comment by SiriusStarr** [comment]
Hm, so I'm confused.  Where is the type error coming from?  I get that only one of the two errors is being preserved, but I don't understand why there are *two* error messages.

**Comment by Gabriel439** [comment]
@SiriusStarr: The logic for deriving `FromDhall` for a union internally uses `typeError` in the base case to signal that a subset of the union does not contain the expected constructor:

https://github.com/dhall-lang/dhall-haskell/blob/0c4b59a02591b65f2076b801e9b5a11d2b2e5596/dhall/src/Dhall.hs#L1464-L1483

**Comment by SiriusStarr** [comment]
But why is it lacking the expected constructor when the ADT is correct and the only thing failing validation is the `extractError`?

**Comment by Gabriel439** [comment]
@SiriusStarr: Because in Haskell derived instances using "GHC generics" have to be defined in a nested way.  See the documentation for: https://hackage.haskell.org/package/base/docs/GHC-Generics.html

Basically the way it works is that GHC takes the Haskell type:

```haskell
data Foo
  = A (Set ())
  | B
  | C
  | D
  deriving (Generic, Show)
```

... and then converts it to `Rep Foo`, which is a type that has a similar shape but more metadata associated with it, like this:

```haskell
>>> :kind! (Rep Foo)
(Rep Foo) :: * -> *
= D1
    ('MetaData "Foo" "Main" "main" 'False)
    ((C1
        ('MetaCons "A" 'PrefixI 'False)
        (S1
           ('MetaSel
              'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
           (Rec0 (Set ())))
      :+: C1 ('MetaCons "B" 'PrefixI 'False) U1)
     :+: (C1 ('MetaCons "C" 'PrefixI 'False) U1
          :+: C1 ('MetaCons "D" 'PrefixI 'False) U1))
```

This `Rep Foo` type is nested, which is more apparent if you elide a few details to look at the overall shape:

```haskell
= D1 … ((C1 … "A" … :+: C1 … "B" …) :+: (C1 … "C" … :+: C1 … "D" …))
```

... so the logic for extracting a Haskell value from the Dhall expression has to be defined in a nested way because the shape of the `Rep Foo` type is defined in a nested way.  This is why you see weird things like it splitting up the union type in half and trying to decode with each half respectively because the `Rep Foo` type is split up in the same way.

**Comment by SiriusStarr** [comment]
Thanks for the explanation.  I am clearly dense because I still don't exactly get how to fix this, but I will keep beating my head against it.  Presumably the answer is don't use generics.

**Comment by Gabriel439** [comment]
@SiriusStarr: No, I think this is still possible using GHC generics.  The problem isn't so much the use of generics, but rather the `ExtractErrors`/`ExtractError` type.  The current types do not permit a way to distinguish between type errors due to custom decoders vs. type errors created by the Generics-based internal logic, so it has to fall back to showing both.

**Comment by akrmn** [comment]
@SiriusStarr, @Gabriel439, I think another way to solve this without changing the Error types would be to define a new class:

```haskell
class GenericFromDhallUnion f where
  foo :: InputNormalizer -> InterpretOptions -> State Int (UnionDecoder (f a))
```

there would be a single instance `GenericFromDhall` for `(:+:)`, which would use `union` on the result of `foo`. A similar change would be needed for `GenericToDhall`. I haven't tried this out yet, though I believe `aeson` does something similar with its GHC generics instances.

**Comment by SiriusStarr** [comment]
@Gabriel439 My apologies, I was ambiguous in saying "I don't understand how to fix this."  I don't mean improving `dhall-haskell` itself (though obviously that is a worthy discussion to have).  I meant I don't know how to circumvent this issue in my own code at this moment, where this is blocking, other than presumably not using Generics?

**Comment by Gabriel439** [comment]
@SiriusStarr: That's correct.  You would have to not use Generics if you want to work around this.  The `Dhall` module provides utilities for manually authoring `Decoder`s for `Union`s if that helps:

https://hackage.haskell.org/package/dhall-1.31.0/docs/Dhall.html#t:UnionDecoder

**Comment by SiriusStarr** [comment]
Yep, I saw thanks, just haven't tackled it yet, haha.

I'm assuming this issue should be left open for the moment, since it represents a non-trivial issue with the state of `dhall-haskell` itself?

**Comment by Gabriel439** [comment]
@SiriusStarr: Yeah, I'm leaving this issue open

**Comment by Gabriel439** [comment]
This is fixed by #1725
